frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+		/// Each ticket's owner.
+		///
+		/// May have residual storage from previous lotteries. Use `TicketsCount` to see which ones
+		/// are actually valid ticket mappings.
+		Tickets: map hasher(twox_64_concat) Index => Option<T::AccountId>;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as frame_system::Config>::AccountId,
+		Balance = BalanceOf<T>,
+		Call = <T as Config>::Call,
+	{
+		/// A lottery has been started!
+		LotteryStarted,
+		/// A winner has been chosen!
+		Winner(AccountId, Balance),
+		/// A ticket has been bought!
+		TicketBought(AccountId, Call),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// An overflow has occurred.
+		Overflow,
+		/// A lottery has not been configured.
+		NotConfigured,
+		/// A lottery has not started.
+		NotStarted,
+		/// A lottery is already in progress.
+		InProgress,
+		/// A lottery has already ended.
+		AlreadyEnded,
+		/// The call is not valid for an open lottery.
+		InvalidCall,
+		/// You are already participating in the lottery with this call.
+		AlreadyParticipating,
+		/// Too many calls for a single lottery.
+		TooManyCalls,
+		/// Failed to encode calls
+		EncodingFailed,
+	}
+}
+
+#[derive(Encode, Decode, Default, RuntimeDebug)]
+pub struct LotteryConfig<BlockNumber, Balance> {
+	/// Index of this lottery
+	index: Index,
+	/// Price per entry.
+	price: Balance,
+	/// Starting block of the lottery.
+	start: BlockNumber,
+	/// Ending block of the lottery.
+	end: BlockNumber,
+	/// Payout block of the lottery.
+	payout: BlockNumber,
+	/// Calls available this lottery.
+	calls: Vec<CallIndex>,
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin, system = frame_system {
+		/// TODO: Expose Constants
+
+		fn deposit_event() = default;
+
+		#[weight = T::WeightInfo::setup_lottery(calls.len() as u32)]
+		fn setup_lottery(
+			origin,
+			price: BalanceOf<T>,
+			start: T::BlockNumber,
+			end: T::BlockNumber,
+			payout: T::BlockNumber,
+			calls: Vec<<T as Config>::Call>,
+		) {
+			T::ManagerOrigin::ensure_origin(origin)?;
+			ensure!(calls.len() <= T::MaxCalls::get(), Error::<T>::TooManyCalls);
+			let indices = Self::calls_to_indices(&calls)?;
+			Lottery::<T>::try_mutate(|lottery| -> DispatchResult {
+				ensure!(lottery.is_none(), Error::<T>::InProgress);
+				let index = LotteryIndex::get();
+				let new_index = index.checked_add(1).ok_or(Error::<T>::Overflow)?;
+				// Use new_index to more easily track everything with the current state.
+				*lottery = Some(LotteryConfig { index: new_index, price, start, end, payout, calls: indices });
+				LotteryIndex::put(new_index);
+				Ok(())
+			})?;
+			Self::deposit_event(RawEvent::LotteryStarted);
+		}
+
+		#[weight =
+			T::WeightInfo::buy_ticket()
+				.saturating_add(call.get_dispatch_info().weight)
+		]
+		fn buy_ticket(origin, call: Box<<T as Config>::Call>) {
+			let caller = ensure_signed(origin.clone())?;
+			call.clone().dispatch(origin).map_err(|e| e.error)?;
```

Comment:
you can avoid this clone by getting call index first, and make `do_buy_ticket` takes call index instead of call